File: articles/systems-thinking-and-levels-of-analysis/python/intervention_level_comparison.py
"""Compare intervention scenarios at different levels of analysis."""

from __future__ import annotations

import csv
from pathlib import Path

DATA = Path(__file__).resolve().parents[1] / "data" / "synthetic_scenarios.csv"


def score_scenario(row: dict[str, str]) -> float:
    individual = float(row["individual_support"])
    organization = float(row["organizational_capacity"])
    institution = float(row["institutional_reform"])
    network = float(row["network_redundancy"])
    ecological_stress = float(row["ecological_stress"])
    return round((individual + organization + institution + network) / ecological_stress, 3)


def main() -> None:
    with DATA.open(newline="", encoding="utf-8") as handle:
        print("scenario_id,scenario_name,cross_scale_score")
        for row in csv.DictReader(handle):
            print(f"{row['scenario_id']},{row['scenario_name']},{score_scenario(row)}")


if __name__ == "__main__":
    main()

File: articles/systems-thinking-and-levels-of-analysis/r/intervention_level_comparison.R
# Intervention-level comparison

article_dir <- normalizePath(file.path(getwd(), "articles", "systems-thinking-and-levels-of-analysis"), mustWork = FALSE)
if (!dir.exists(article_dir)) article_dir <- normalizePath(file.path(getwd(), ".."), mustWork = FALSE)

scenarios <- read.csv(file.path(article_dir, "data", "synthetic_scenarios.csv"))
scenarios$cross_scale_score <- with(
  scenarios,
  (individual_support + organizational_capacity + institutional_reform + network_redundancy) / ecological_stress
)

print(scenarios[, c("scenario_id", "scenario_name", "cross_scale_score")])
